lib/Service/LockService.php

```diff
@@ -132,6 +132,7 @@ public function getLockForNodeIds(array $nodeIds): array {
 				$locks[$nodeId] = $this->lockCache[$nodeId];
 			} else {
 				$locksToRequest[] = $nodeId;
+				$this->lockCache[$nodeId] = false;
```

Comment:
This is the core of this change, right? The rest is an independent performance tweak.

I'm not sure I understand the context well enough yet.

Does `getLockForNodeIds` lock the node ids in question or is this just for retrieving the lock records?

This line will have the effect that `$this->lockCache[$nodeId]` is `false`
until it's updated in the `foreach` loop below based on the results of `$this->locksRequest->getFromFileIds($fileIds)`.

If this function is called again in the meantime `$locks[$nodeId]` will be set to `false` as well as the `if` condition in line 131 will evaluate to true for the `nodeId` in question.
Do we need to think about that or is `$this->lockCache` only available in a synchronous context?
Is this the desired effect?
It will make sure the `nodeId` does not end up in `$locksToRequest` - so fewer locks will be requested. But can we be sure that this is not just an intermediate value until it's overwritten in the foreach loop?

Reply:
> Does getLockForNodeIds lock the node ids in question or is this just for retrieving the lock records?

It only retrieves and is used to prefetch the lock state in the propfind dav response to avoid that each file is calling getLockForNodeId individually.

> Do we need to think about that or is $this->lockCache only available in a synchronous context?
> Is this the desired effect?
> It will make sure the nodeId does not end up in $locksToRequest - so fewer locks will be requested. But can we be sure that this is not just an intermediate value until it's overwritten in the foreach loop?

For files with locks the false is expected to be an intermediate value, however for other files that are not locked, the false value will lead to the lock state being considered cached if follow up method calls try to check the lock.

Reply:
> For files with locks the false is expected to be an intermediate value, however for other files that are not locked, the false value will lead to the lock state being considered cached if follow up method calls try to check the lock.

So these follow up method calls try to check the lock... the lock state is considered to be cached - so they do not trigger additional requests i assume. And then... how do they know weather the file is locked or not? Do they wait for the intermediate value of `false` to be resolved or do they assume the file to be (not) locked?

Reply:
If the file is locked the false will be overwritten by https://github.com/nextcloud/files_lock/pull/196/files#diff-cbd147b6321b8e183a29abf15821d7989b15259d774fa1f2c391618e34deb10dR149-R153 in this method here, if not false is kept as the final value to indicate the file is **not** locked.

Reply:
ahh.. now i got it.

Reply:
Yes, also the cache is only for the current request in memory, nothing persisted across requests.